src/InteractiveWindow/Editor/InteractiveWindow.cs

```diff
@@ -1447,88 +1461,126 @@ private void MeasurePrompts(int startLine, int endLine, out int minPromptLength,
+                });
+
+            if (index < 0)
             {
-                int comp = x.LineNumber - y.LineNumber;
-                return comp != 0
-                    ? comp
-                    : GetKindOrdinal(x) - GetKindOrdinal(y);
+                Debug.Assert(~index == sourceSpans.Count);
+                Debug.Assert(point.Position == point.Snapshot.Length);
+                index = ~index - 1;
             }
 
-            /// <summary>
-            /// Within a given line, the order is Output &lt; Prompt &lt; Language/StdIn.
-            /// </summary>
-            /// <remarks>
-            /// While, visually, it appears that every line begins with a prompt, a line
-            /// may actually begin with an empty output prompt, because at any given time
-            /// there must be a writable output buffer for <see cref="AppendOutput"/>.
-            /// Generally, this occurs when a submission has no output.
-            /// </remarks>
-            private static int GetKindOrdinal(ReplSpan span)
+            Debug.Assert(index >= 0);
+            Debug.Assert(index < sourceSpans.Count);
+
+            // Find the nearest preceding prompt.
+            while (!IsPrompt(sourceSpans[index].Snapshot))
             {
-                switch (span.Kind)
-                {
-                    case ReplSpanKind.Output:
-                        return 1;
-                    case ReplSpanKind.Prompt:
-                    case ReplSpanKind.SecondaryPrompt:
-                    case ReplSpanKind.StandardInputPrompt:
-                        return 2;
-                    case ReplSpanKind.Language:
-                    case ReplSpanKind.StandardInput:
-                        return 3;
-                    default:
-                        throw new InvalidOperationException($"Unexpected {nameof(ReplSpanKind)} '{span.Kind}'");
-                }
+                index--;
             }
-        }
 
-        /// <summary>
-        /// Creates the language span for the last line of the active input.  This span
-        /// is effectively edge inclusive so it will grow as the user types at the end.
-        /// </summary>
-        private CustomTrackingSpan CreateLanguageTrackingSpan(Span span)
-        {
-            return new CustomTrackingSpan(
-                _currentLanguageBuffer.CurrentSnapshot,
-                span,
-                PointTrackingMode.Negative,
-                PointTrackingMode.Positive);
+            return index;
         }
 
-        /// <summary>
-        /// Creates the tracking span for a line previous in the input.  This span
-        /// is negative tracking on the end so when the user types at the beginning of
-        /// the next line we don't grow with the change.
-        /// </summary>
-        private CustomTrackingSpan CreateNonGrowingLanguageTrackingSpan(Span span)
+        private static int BinarySearch<T>(ReadOnlyCollection<T> collection, Func<T, int> compare)
```

Comment:
We could use List<T>.BinarySearch but that method expects to search for a T (in this case, a SnapshotSpan) so we'd need to create a fake SnapshotSpan for the point and have a custom IComparer<SnapshotSpan> that expects one argument to be the point.